lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp

```diff
@@ -94,6 +95,10 @@ class TSNodeLowering : public NodeLowering {
         // Only used from bmm currently.
         return InferBmm(node);
       }
+      case at::aten::cat: {
```

Comment:
This is helpful in understanding the "what does a vendor have to do to lower an op".

The part that I already knew was that you need to update the yaml to say "I'm lowering this op". It looks like in addition to adding a new kernel (which will fail at compile time if you forget to include), here you also need to provide:

(a) a shape inference function
(b) a node lowering

And you'll fail at runtime if you forget either one.

Two followup Q's related to this:

(a) Just thinking out loud, I wonder how feasible it is to turn the runtime errors into compile-time errors. Maybe codegen could help there (we generate node lowering declarations for every op you promise to provide one for)

(b) We talked a little before about LTC doing the shape-checking automatically for some ops, and providing multiple API's - one that assumes LTC does the shape checking, and one that expects the vendor to do the shape checking. What are your thoughts on the timeline for that? It seems like it'd be useful even now, since you have to provide the shape inference formulas anyway in order to get the TS backend to work.

Reply:
(a) Yeah, codegen is the only way I could see this happen. You'd need meta tensor coverage to be able to codegen inference, but it's definitely doable.
(b) Vendors which implement XLA lowering can simply use it to trivially provide inference, vendors which have TorchScript backends can use ts_backend. I don't know if inference helpers are truly needed, from a pragmatic point of view. We can go without, especially with the meta tensor providing the inference functionality.

Reply:
For (a): Just making sure I have the same picture in my head as you - I could imagine using codegen without requiring full meta tensor coverage, right? The codegen is aware of which ops have meta tensor support (basically just "which kernels are structured"), so it could codegen meta API calls for what's available, and expect (and generate declarations for) shape inference functions for the other ops.

For (b): It sounds like you're saying most vendors will integrate either with TorchScript or by writing lowerings from XLA, but not by writing lowerings from LTC directly. Is that right?

Reply:
Yes, you could have some naming convention instead of using meta tensor everywhere and codegen to that. I just think meta tensors would be cleaner and would prefer to have one mechanism than two (and relying on naming conventions).

For your second question, I suspect we'll see some lowering from LTC directly, but by the time that happens I'd expect our builtin inference support to be robust. I'm just making the argument that we cover current targets well - people we talked to do one of these two.

Reply:
> I just think meta tensors would be cleaner and would prefer to have one mechanism than two

>  I suspect we'll see some lowering from LTC directly, but by the time that happens I'd expect our builtin inference support to be robust.

My main worry is that there's a long tail of LTC-enabled ops that currently don't have a meta API (on the order of 200-300), so this might not happen for a while if we're going to require 100% coverage before switching to the meta API.

I could imagine some alternatives that are less robust than codegen but easier to build. For example, right now vendors find out that they're missing shape inference functions at runtime. You already have shape inference code written for all of the ops that the TS backend supports - just from me looking over the code, is it feasible to move the shape inference into each LTC node's constructor? For any ops that we don't already have shape inference support, we can fall back to the vendor-defined `Infer()` function. The downside is that vendors don't have a clear signal of which ops they need to define their own shape inference for. It seems like that's already the case though - for example, I see that all of the Unary op nodes don't call `Infer()`, and instead just take the input's shape to be the same as the output shape.



Hmm okay, my understanding is that each LTC-enabled operator has its own `Node` class, and they all perform shape inference by calling into ``so I can imagine 3 ways of doing this (and let me know if you agree with this picture):

(1) Current state: the vendor provides their own giant `NodeLowering::Infer()` implementation, which is a big switch statement that provides shape inference for each op. I see two downsides:
- Everyone who integrates with LTC needs to provide their own complete set of shape inference functions. That also means that you have to duplicate the existing shape inference code between the TS and XLA plugins
- Missing shape inference functions aren't caught until runtime

(2)

Reply:
We don't need 100% coverage, we just need coverage for operators we cared about in PT/XLA. The rest go on the interop path which doesn't care about explicit inference support anyway.

You're right that we can just make the inference functionality we've implemented part of node constructors, I plan to do that en-masse in a subsequent commit.